overlord/fdestate/handlers.go

```diff
@@ -19,47 +19,177 @@
+	for _, container := range containers {
+		containerDevicePath[container.ContainerRole()] = container.DevPath()
+	}
+
+	// IMPORTANT: this clean up must be decalred as early as possible
+	// to account for real errors and potential re-runs (that will fail
+	// anyway as soon as the recovery key ID is reused).
+	defer func() {
+		if err == nil {
+			return
+		}
+		for _, keyslotRef := range keyslotRefs {
+			devicePath := containerDevicePath[keyslotRef.ContainerRole]
+			if err := secbootDeleteContainerKey(devicePath, keyslotRef.Name); err != nil {
+				// best effort deletion, log errors only
+				logger.Noticef("failed to delete %s during clean up: %v", keyslotRef.String(), err)
+			}
+		}
+	}()
+
+	currentKeyslots, missing, err := m.GetKeyslots(keyslotRefs)
+	if err != nil {
+		return fmt.Errorf("failed to find key slots: %v", err)
+	}
+	if len(missing) == 0 {
+		// this could be re-run and all key slots were already added, do nothing
+		return nil
+	}
+	if len(currentKeyslots) != 0 {
+		return &keyslotsAlreadyExistsError{keyslots: currentKeyslots}
+	}
+
+	rkey, err := m.getRecoveryKey(recoveryKeyID)
+	if err != nil {
+		// most likely a re-run as keys expire after first use and
+		// clean up is needed.
+		return fmt.Errorf("failed to find recovery key with id %q: %v", recoveryKeyID, err)
+	}
+
+	for _, keyslotRef := range keyslotRefs {
+		devicePath := containerDevicePath[keyslotRef.ContainerRole]
+		if err := secbootAddContainerRecoveryKey(devicePath, keyslotRef.Name, rkey); err != nil {
+			return fmt.Errorf("failed to add recovery key slot %s: %v", keyslotRef.String(), err)
+		}
+	}
+
 	return nil
 }
 
 func (m *FDEManager) doRemoveKeys(t *state.Task, tomb *tomb.Tomb) error {
-	// TODO:FDEM: implement recovery key removal, this is currently only a
-	// mock task for testing.
-
-	// TODO:FDEM:
-	//   - this might be a re-run, make task idempotent to be reselient to
-	//     abrupt reboot/shutdown.
-	//   - distinguish between errors (undo) and pure-reboots (re-run).
-	//   - conflict detection for key slot tasks is important because it
-	//     reduces the possible states we could end up in.
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	var keyslotRefs []KeyslotRef
+	if err := t.Get("keyslots", &keyslotRefs); err != nil {
+		return err
+	}
+
+	// XXX: unlock state and let conflict detection handle the rest?
+
+	// we only care about current key slots because this might be
+	// a re-run due a force reboot or abrupt shutdown, so we want
+	// to continue deleting the remaining key slots.
+	currentKeyslots, _, err := m.GetKeyslots(keyslotRefs)
+	if err != nil {
+		return fmt.Errorf("failed to find key slots: %v", err)
+	}
+
+	for _, keyslot := range currentKeyslots {
+		if err := secbootDeleteContainerKey(keyslot.devPath, keyslot.Name); err != nil {
+			// XXX: keep going and report errors afterwards?
+			return fmt.Errorf("failed to remove key slot %s: %v", keyslot.Ref().String(), err)
```

Comment:
to be idempotent we should probably only log if the error is that the keyslot doesn't exist anymore

Reply:
Doc comment for `GetKeyslots` could definitely be improved.